NFT/src/main.sw

```diff
@@ -0,0 +1,359 @@
+contract;
+
+dep contract_abi;
+dep data_structures;
+dep errors;
+dep events;
+dep utils;
+
+use contract_abi::NFT;
+use data_structures::MetaData;
+use errors::{AccessError, ApprovalError, InitError, InputError};
+use events::{ApprovalEvent, BurnEvent, MintEvent, OperatorEvent, TransferEvent};
+use utils::token_metadata;
+
+use std::{
+    assert::require,
+    chain::auth::{AuthError, msg_sender},
+    identity::Identity,
+    logging::log,
+    option::Option,
+    result::Result,
+    storage::StorageMap,
+    vec::Vec,
+};
+
+storage {
+    /// Determines if only the `admin` is allowed to call the mint function.
+    /// This is only set on the initalization of the contract.
+    access_control: bool,
+
+    /// Stores an `Option` of the `Identity` that has permission to mint if `access_control` is set to true.
+    /// Will store `None` if this contract does not have `access_control` set.
+    /// Only the `admin` is allowed to change the `admin` of the contract.
+    admin: Option<Identity>,
+
+    /// Used of O(1) lookup of the number of tokens owned by each `Identity`.
+    /// This increments or decrements when minting, transfering ownership, and burning tokens.
+    /// Map(Identity => balance)
+    balances: StorageMap<Identity,
+    u64>, /// Stores the `Metadata` for each token based on the token's `u64` id
+    /// Map(token_id => Metadata)
+    meta_data: StorageMap<u64,
+    Option<MetaData>>, /// Maps a tuple of (owner, operator) identities and stores whether the
+    /// operator is allowed to transfer ALL tokens on the owner's behalf.
+    /// Map((owner, operator) => approved)
+    operator_approval: StorageMap<(Identity,
+    Identity), bool>, /// The total number of tokens that have been minted.
+    /// This should only be incremented.
+    token_count: u64,
+
+    /// The total supply tokens that can be minted. Can only be set on the initalization of the
+    /// contract.
+    token_supply: u64,
```

Comment:
I feel there's something lacking here. 
First, I don't think ? there's any storage variable tracking the _total current_ tokens in existence?
I might have missed something as it very early, but I might suggest something like:
```
supply_cap: u64; // The total supply tokens that can be minted.

total_supply: u64 // The number of tokens currently in existence, Incremented on mint, decremented on burn.

tokens_ever_minted: u64; // The number of tokens that have been minted.
```